ARCHITECTURE.md

```diff
@@ -0,0 +1,162 @@
+
+Preprocessing also renames methods (since dart doesn't support overloading). Ad-hoc overloading can lead to a methods signature being different than the same-named method in its superclass - which is an error. We actually do some circus to prevent this. We save java signature of each method as a string, which is a unique representation. Now it's possible to have a map per class of these methods and their associated disambiguating numbers. By ensuring the superclass is sorted before subclass, we ensure methods with same signature get the same number. (Sounds like topological sort, but actually it's not needed. We just lazily preprocess the parent and mark it as preprocessed.)
+
+### File mapping
+Depending on `output` >> `dart` >> `structure` key in config, output can be package structured or single file. In package structure, the original source is mirrored, with each top level class being mapped to a separate dart file. An `_init.dart` file will be created to hold common variables. In each package, `_package.dart` file will be created, which exports all classes directly (__not transitively__) in the package. This is equivalent of the wildcard import in Java.
+
+Note that even in package structure mode, nested classes will be in the same file as their parent class, and not separate file. Nested classes will be renamed using an underscore. `A$B` becomes `A_B`.
+
+In single file mode, all classes will be written to a single dart file, and renamed if required.
+
+There are 2 separate writer classes ([FilesWriter](jnigen/lib/src/writers/files_writer.dart) and [SingleFileWriter](jnigen/lib/src/single_file_writer.dart)) for each strategy. They descend from same abstract class and share some code.
+
+Similarly, for both binding types, (`c_based` and `pure_dart`), there are `BindingsGenerator` subclasses which provide an identical interface. The pre and post import boilerplates differ based on binding type, which is provided by overridden methods.
+
+### Resolution
+A method signature, or a class declaration (in form of an `extends` clause), can refer to other classes. If the class is included in the same `jnigen` config, the generated code should refer to that class. Otherwise it should fall back to `JniObject`.
+
+Resolution process is different for single file and package structured bindings. In case of single file bindings, using the final name of the referred class is sufficient. In package structured bindings, the following is an abstract description of algorithm is used to resolve shortest path to class B from class A.
```

Comment:
Resolution -> The resolution
single file -> single-file
package structured -> package-structured